### PR TITLE
Billing credit card flow

### LINKED ISSUE #1757: Upgrade Plan Flow - Credit Card
Overview of the flow being implemented is on [Figma](https://www.figma.com/file/wRMGqc11iQLVpDVmqXVXWK/ChainSafe-Files?node-id=2192%3A1309) - Designs are now mostly complete for this.

image

- Add a Subscription Overview modal to be displayed after the user has selected a new subscription. This should allow a user to choose annual or monthly billing
image

- Add a payment method selection modal to allow a user to select a payment method for their subscription. 
  - ~~The crypto option should only be enabled if `Annual` subscription was selected previously.~~ For now just have the crypto option disabled
  - The card option should only be enabled if a card has been added
  - If a card has been added, display the card details
image
image

- If the user does not have a card added an expansion panel with the card details form should be displayed allowing the user to add a card.
  - Handle 3D Secure authentication while adding a card. When `Use this payment method` is clicked, display the 3DS iframe over the Payment Method modal:
image

- Add a confirmation modal once successful payment has been processed
image

### Questions
- [x] How to integrate the crypto into the existing UX
  - Crypto will be handled in the next issue
- [x] Can we use the existing Add Card component in this modal
  - Yes
- [x] Does choosing to `Edit` an existing card just replace the default card on Stripe?
  - Currently no, and we should figure out where this functionality will be implemented (API or UI) as with the current UI it makes sense not to store any unnecessary

### PR DESCRIPTION
**Pull request by tanmoyAtb**
closes #1757 

API updates are now in.

Submission checklist: 

- [x] Select plan page
- [x] plan details page
- - [x] monthly
- - [x] yearly

- [x] Card or wallet page
- [x] Add card functionality
- [x] payment page
- [x] Plan change success page

- [x] Layout
- - [x] Change looks good in the desktop web ui
- - [x] Change looks good in the mobile web ui

- [x] Theme
- - [x] Components / elements inspected in light mode 
- - [x] Components / elements inspected in dark mode

### CODE REVIEW

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/AddCard/AddCard.tsx [hunk 1]
```diff
@@ -160,93 +136,72 @@ const AddCardModal = ({ isModalOpen, onClose }: IAddCardModalProps) => {
...
-              component="h4"
-            >
-              <Trans>Add a credit card</Trans>
-            </Typography>
-          </Grid>
-          <CardNumberElement
+    <form onSubmit={handleSubmitPaymentMethod}>
+      <div
+        className={classes.root}
+        data-cy="modal-add-card"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/AddCard/AddCard.tsx]: That's not a modal any more. 
```suggestion
        data-cy="form-add-card"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/ChangePlanModal.tsx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  onClose: () => void
+}
+
+const ChangeProductModal = ({ onClose }: IChangeProductModal) => {
+  const classes = useStyles()
+  const { getAvailablePlans, changeSubscription } = useBilling()
+  const [selectedPlan, setSelectedPlan] = useState<Product | undefined>()
+  const [selectedPrice, setSelectedPrice] = useState<ProductPrice | undefined>()
+  const [slide, setSlide] = useState<
+  "select"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/ChangePlanModal.tsx]: nit: Can we extract this in a type and use it here.

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/ChangePlanModal.tsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      active={true}
+      maxWidth="md"
+      className={classes.root}
+      injectedClass={{
+        inner: classes.inner
+      }}
+      testId="change-product"
+    >
+      {
+        slide === "select" ? <SelectPlan
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/ChangePlanModal.tsx]: This avalanche of ternaries are super hard to diggest.
Can we rather use something like 
```
{slide === "xyz && (
    <Component
        prop1="prop"
        prop2="prop"
    />
})
```?

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+          onClick: goToSelectPlan
+        }, {
+          text: t`Plan details`,
+          onClick: goToPlanDetails
+        }, {
+          text: t`Payment method`
+        }]}
+        hideHome={true}
+      />
+      <Typography variant="h5"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx]: next line
```suggestion
      <Typography
        variant="h5"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        component="h4"
+        className={classes.heading}
+      >
+        <Trans>Select payment method</Trans>
+      </Typography>
+      <Typography variant="body1"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx]: next line
```suggestion
      <Typography
        variant="body1"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            label="USDC, BTC, or ETH (Annual only)"
+            value="orange"
+            onChange={() => setPaymentMethod("crypto")}
+            checked={paymentMethod === "crypto"}
+            labelClassName={classes.radioLabel}
+            disabled={true}
+          />
+        </>
+        : <div className={classes.addCardWrapper}>
+          <AddCard submitText={t`Use this card`}
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx]: new line
```suggestion
          <AddCard
            submitText={t`Use this card`}
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            footerClassName={classes.footer}
+            onCardAdd={() => setView("selectPaymentMethod")}
+          />
+        </div>
+      }
+      <Divider className={classes.divider} />
+      <section className={classes.bottomSection}>
+        {view === "addCard" ? <div>
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx]: The ternary doesn't look useful here, can we use `condition && <Component/>`?

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+            onClick={() => setView("selectPaymentMethod")}
+            className={classes.linkButton}
+          >
+            <Trans>Go  back</Trans>
+          </Typography>
+        </div> : <div />
+        }
+        <div className={classes.buttons}>
+          <Button
+            onClick={() => onClose()}
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx]: onClose
```suggestion
            onClick={onClose}
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+const PlanSuccess = ({
+  plan,
+  onClose,
+  planPrice
+}: IPlanSuccess) => {
+  const classes = useStyles()
+
+  return (
+    <article className={classes.root}>
+      <Typography variant="body1"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx]: new line
```suggestion
      <Typography
        variant="body1"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        component="p"
+        className={classes.headingBadge}
+      >
+        <Trans>Confirmation</Trans>
+      </Typography>
+      <div className={clsx(classes.middleRowBox, classes.headingBox)}>
+        <Typography variant="h5"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx]: new line
```suggestion
        <Typography
          variant="h5"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+          component="h4"
+        >
+          <Trans>Plan changed successfully</Trans>
+          &nbsp;&nbsp;
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx]: Hmm this is inelegant :D Can we please use css?

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+            <CheckIcon className={classes.tickIcon} />
+            <Typography component="p"
+              variant="body1">
+              {plan.description}
+            </Typography>
+          </div>
+        </div>
+      </div>
+      <div className={classes.rowBox}>
+        <Typography component="p"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx]: new line
```suggestion
        <Typography
          component="p"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+            className={classes.textLink}
+          >
+            <Trans>invoices here</Trans>
+          </Link>
+        </Typography>
+      </div>
+      <section className={classes.bottomSection}>
+        <div className={classes.buttons}>
+          <Button
+            onClick={() => onClose()}
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PlanSuccess.tsx]: onClose
```suggestion
            onClick={onClose}
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+            <Typography
+              component="p"
+              variant="body1"
+              className={classes.planTitle}
+            >
+              {plan.name}
+            </Typography>
+            {monthly && <Typography component="h4"
+              variant="h4">
+              {monthly.unit_amount ? <>
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx]: Can we pleae use `?` and `:` at the beginning of the line?

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+            </div>
+          </div>
+        })
+        }
+      </section>
+      <section className={classes.bottomSection}>
+        {desktop &&
+        <a
+          className={classes.link}
+          href="http://chainsafe.io"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx]: Can we use the Route component?

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+            <Trans>
+              Not sure what to pick? Learn more about our plans
+            </Trans>
+          </Typography>
+          <ExternalSvg />
+        </a>
+        }
+        <div className={classes.buttons}>
+          <Button
+            onClick={() => onClose()}
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx]: onClose
```suggestion
            onClick={onClose}
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/ConfirmPlan.tsx [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+          variant="body1"
+          className={classes.error}
+        >
+          <Trans>Failed to change subscription</Trans>
+        </Typography>
+      }
+      <section className={classes.bottomSection}>
+        <div className={classes.buttons}>
+          <Button
+            onClick={() => onClose()}
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/ConfirmPlan.tsx]:
```suggestion
            onClick={onClose}
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+          <AddCard submitText={t`Use this card`}
+            footerClassName={classes.footer}
+            onCardAdd={() => setView("selectPaymentMethod")}
+          />
+        </div>
+      }
+      <Divider className={classes.divider} />
+      <section className={classes.bottomSection}>
+        {view === "addCard" ? <div>
+          <Typography variant="body1"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx]:
```suggestion
          <Typography
            variant="body1"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+              {
+                monthly?.metadata?.storage_size_bytes
+                  ? <>
+                    <b>{formatBytes(Number(monthly?.metadata?.storage_size_bytes), 2)}</b>&nbsp;
+                    <Trans>of storage</Trans >
+                  </>
+                  : plan.description
+              }
+            </Typography>
+            <Button variant="primary"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx]:
```suggestion
            <Button
              variant="primary"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+              disabled={!isPlanSelectable}
+              onClick={() => onSelectPlan(plan)}
+            >
+              <Trans>Select plan</Trans>
+            </Button>
+          </div> : <div
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx]: Beginning of line?
```suggestion
          </div> : <div
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+        <div className={classes.buttons}>
+          <Button
+            onClick={() => onClose()}
+            variant="secondary"
+          >
+            <Trans>
+              Cancel
+            </Trans>
+          </Button>
+          {!desktop && <Button variant="primary"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/SelectPlan.tsx]:
```suggestion
          {!desktop && <Button
            variant="primary"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/CurrentPlan.tsx [hunk 22]
```diff
@@ -78,7 +78,8 @@ const CurrentProduct = ({ className }: ICurrentProduct) => {
         >
           {currentSubscription?.product.name}
         </Typography>
-        : <Loading />
+        : <Loading size={36}
+          type="inherit" />
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/CurrentPlan.tsx]:
```suggestion
        : <Loading
          size={36}
          type="inherit"
        />
```

**Review by Tbaut — COMMENTED** [review]
Went through the code and found only some lint things. Will play with now :)

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/AddCard/AddCard.tsx [hunk 23]
```diff
@@ -96,11 +70,13 @@ const useStyles = makeStyles(
...
-  isModalOpen: boolean
-  onClose: () => void
+  submitText: string
+  onCardAdd?: () => void
+  onClose?: () => void
+  footerClassName?: string
 }
 
-const AddCardModal = ({ isModalOpen, onClose }: IAddCardModalProps) => {
+const AddCardModal = ({ onCardAdd, submitText, onClose, footerClassName }: IAddCardModalProps) => {
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/AddCard/AddCard.tsx]:
```suggestion
const AddCard = ({ onCardAdd, submitText, onClose, footerClassName }: IAddCardModalProps) => {
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/AddCard/AddCard.tsx [hunk 24]
```diff
@@ -258,21 +213,21 @@ const AddCardModal = ({ isModalOpen, onClose }: IAddCardModalProps) => {
...
+            loading={loadingPaymentMethodAdd}
+            disabled={loadingPaymentMethodAdd}
+          >
+            {submitText}
+          </Button>
+        </Grid>
+      </div>
+    </form>
   )
 }
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/AddCard/AddCard.tsx:259]: The line below needs to change for AddCard, just I can't edit here.

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+      <Divider className={classes.divider} />
+      {view === "selectPaymentMethod"
+        ? <>
+          <div className={classes.rowBox}>
+            <RadioInput
+              label={defaultCard ? `•••• •••• •••• ${defaultCard.last_four_digit}` : "Credit card" }
+              value="creditCard"
+              onChange={() => setPaymentMethod("creditCard")}
+              checked={paymentMethod === "creditCard"}
+              labelClassName={classes.radioLabel}
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionTab/ChangePlan/PaymentMethod.tsx]: This should be disabled if there's no default card. I was able to change plan without card.
```suggestion
              labelClassName={classes.radioLabel}
              disabled={!defaultCard}
```

**Review by Tbaut — COMMENTED** [review]
:tada: It's wired very well, awesome work Tanmoy! 

a couple remarks more on the design that the implementation itself
I found this drop-down a little surprising since I've never seen anything similar anywhere else.
Also I think the / instead of > from the original design does make a visual difference for the breadcrumb.
image


Should we place this "go back" next to the "use this card" button, because it's effectively canceling the "add card" process
image

**Comment by tanmoyAtb** [comment]
> 🎉 It's wired very well, awesome work Tanmoy!
> 
> a couple remarks more on the design that the implementation itself I found this drop-down a little surprising since I've never seen anything similar anywhere else. Also I think the / instead of > from the original design does make a visual difference for the breadcrumb. image
> 
> Should we place this "go back" next to the "use this card" button, because it's effectively canceling the "add card" process image

Hey, I made the changes to the Go back button, makes sense there.

So we already have these breadcrumbs in the file browsers. If you keep moving inside folders after folders, you will see these breadcrumbs in mobile screen. I did not want to add new design of breadcrumbs here. We should stick to one.

What we have in browsers:
 
Screenshot 2021-12-08 at 11 30 31 AM

Screenshot 2021-12-08 at 11 31 45 AM

**Review by Tbaut — APPROVED** [review]
Pushed plenty of small things to prevent back & forth for the review. Looking great :tada:

**Review by FSM1 — APPROVED** [review]

**Pull request merged**